core/matrix/csr.cpp

```diff
@@ -520,123 +527,270 @@ std::unique_ptr<LinOp> Csr<ValueType, IndexType>::conj_transpose() const
 }
 
 
+template <typename ValueType, typename IndexType>
+std::unique_ptr<Csr<ValueType, IndexType>> Csr<ValueType, IndexType>::permute(
+    ptr_param<const Permutation<IndexType>> permutation,
+    permute_mode mode) const
+{
+    const auto exec = this->get_executor();
+    const auto size = this->get_size();
+    const auto nnz = this->get_num_stored_elements();
+    if ((mode & permute_mode::symmetric) == permute_mode::none) {
+        return this->clone();
+    }
+    if ((mode & permute_mode::symmetric) == permute_mode::symmetric) {
+        GKO_ASSERT_IS_SQUARE_MATRIX(this);
+    }
+    if ((mode & permute_mode::rows) == permute_mode::rows) {
+        GKO_ASSERT_EQ(size[0], permutation->get_size()[0]);
+    }
+    if ((mode & permute_mode::columns) == permute_mode::columns) {
+        GKO_ASSERT_EQ(size[1], permutation->get_size()[0]);
+    }
+    auto result = Csr::create(exec, size, nnz, this->get_strategy()->copy());
+    auto local_permutation = make_temporary_clone(exec, permutation);
+    std::unique_ptr<const Permutation<IndexType>> inv_permutation;
+    const auto perm_idxs = local_permutation->get_const_permutation();
+    const IndexType* inv_perm_idxs{};
+    // to permute columns, we need to know the inverse permutation
```

Comment:
Maybe also mention why we can not permute the column directly?
permute column (`A'[i, j] = A[i, p[j]]`) directly needs to go through all j in `#cols`  not the sparsity